### PR TITLE
Build: Calypso with calypso-build

### PR DESCRIPTION
**Pull request by ockham**
#### Changes proposed in this Pull Request

Use `calypso-build` to build Calypso. See https://github.com/Automattic/wp-calypso/pull/32630#issuecomment-491720808 for the rationale.

#### Testing instructions

```
yarn distclean && yarn ci
yarn start
```

Does Calypso build and run?

```
CALYPSO_ENV=production yarn build-client-if-prod
```

Does this build? (This takes a while.)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -224,10 +224,10 @@
...
 		"prebuild-server": "mkdirp build",
-		"build-server": "cross-env-shell BROWSERSLIST_ENV=server NODE_PATH=$NODE_PATH:server:client:. webpack --display errors-only --config webpack.config.node.js",
+		"build-server": "cross-env-shell BROWSERSLIST_ENV=server NODE_PATH=$NODE_PATH:server:client:. calypso-build --display errors-only --config webpack.config.node.js",
 		"build-client": "npm run build-client-evergreen",
-		"build-client-fallback": "cross-env-shell BROWSERSLIST_ENV=defaults NODE_PATH=$NODE_PATH:server:client:. node $NODE_ARGS ./node_modules/webpack/bin/webpack.js --display errors-only",
-		"build-client-evergreen": "cross-env-shell BROWSERSLIST_ENV=evergreen NODE_PATH=$NODE_PATH:server:client:. node $NODE_ARGS ./node_modules/webpack/bin/webpack.js --display errors-only",
+		"build-client-fallback": "cross-env-shell BROWSERSLIST_ENV=defaults NODE_PATH=$NODE_PATH:server:client:. node $NODE_ARGS ./packages/calypso-build/bin/calypso-build.js --display errors-only",
+		"build-client-evergreen": "cross-env-shell BROWSERSLIST_ENV=evergreen NODE_PATH=$NODE_PATH:server:client:. node $NODE_ARGS ./packages/calypso-build/bin/calypso-build.js --display errors-only",
```
Comment (ockham) [inline package.json]: The `node $NODE_ARGS ./packages/calypso-build/bin/calypso-build.js` stuff is mildly annoying but apparently required to take ` NODE_ARGS=--max_old_space_size=8192` from [L232](https://github.com/Automattic/wp-calypso/pull/33035/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R232) into account. Ideally, we can figure out why building Calypso in production mode takes up so much memory, fix the issue, and drop this stuff.

**Comment by simison** [comment]
Leaving a note here that once this PR is merged, https://github.com/Automattic/jetpack/pull/12394 can be revisited.

**Comment by sirreal** [comment]
I pushed a small change that I'd be happy to discuss. It uses `./node_modules/.bin/calypso-build` directly and consistently where it's run in `package.json`.

**Review by sirreal — DISMISSED** [review]
In my smoke testing this works well.

I've run e2e, let's wait for the report 🤞

**Comment by jsnajdr** [comment]
@ockham @sirreal This PR is still worth finishing, isn't it? And is doesn't look very complex.

**Comment by sirreal** [comment]
> @ockham @sirreal This PR is still worth finishing, isn't it? And is doesn't look very complex.

This does seem worthwhile. I've rebased it, let's see how e2e goes and get this reviewed.

**Comment by blowery** [comment]
@sirreal Do you recall why this one happened?

> I pushed a small change that I'd be happy to discuss. It uses ./node_modules/.bin/calypso-build directly and consistently where it's run in package.json.

Seems like we should be able to just use `calypso-build` plain since it's being called via an npm script? Did something fail?

**Comment by sirreal** [comment]
> Do you recall why this one happened?
> 
> > I pushed a small change that I'd be happy to discuss. It uses ./node_modules/.bin/calypso-build directly and consistently where it's run in package.json.
> 
> Seems like we should be able to just use `calypso-build` plain since it's being called via an npm script? Did something fail?

I don't recall, we can try rolling it back and see what happens.

**Comment by jsnajdr** [comment]
> Seems like we should be able to just use `calypso-build` plain since it's being called via an npm script? Did something fail?

There's one gotcha when looking up the `node_modules/.bin` binaries: unlike regular `require` or `import` calls, the `node_modules/.bin` lookup doesn't look at the parent directories recursively.

When there is a nested `package.json` in:
```
wp-calypso/packages/foo/package.json
```
that contains a script:
```
{
  "scripts": {
    "build": "calypso-build"
  }
}
```
Then that script runs with `$PATH` containing `wp-calypso/packages/foo/node_modules/.bin`. The `calypso-build` binary must be there to be found. If the binary is in `wp-calypso/node_modules/.bin` then the script won't find it.

`lerna run` apparently does some magic (either with `$PATH` or with `cwd`, not sure) to make it all work with `calypso-build` in the root `node_modules/.bin` and scripts in nested packages.

**Comment by sirreal** [comment]
I've rebased this and it still seems to work fine. Is there interest in moving forward with it?

**Comment by sirreal** [comment]
> @sirreal Do you recall why this one happened?
> 
> > I pushed a small change that I'd be happy to discuss. It uses ./node_modules/.bin/calypso-build directly and consistently where it's run in package.json.
> 
> Seems like we should be able to just use `calypso-build` plain since it's being called via an npm script? Did something fail?

Reviewing the code, the place where `node_modules/.bin/calypso-build` is used is where it's passed directly to node. It makes sense that in this case there's no special resolution to find `calypso-build`: `node ${NODE_ARGS:---max-old-space-size=8192} ./node_modules/.bin/calypso-build`

**Comment by ockham** [comment]
> I've rebased this and it still seems to work fine.

:partying_face: Thanks a lot, Jon!

> Is there interest in moving forward with it?

@Automattic/team-calypso Can I ask y'all to review/approve/reject this? I don't have a lot of time/headspace for Calypso changes these days, so I'd be thankful if you could help me out here :smile:

**Comment by scinos** [comment]
I have mixed feelings about this. Looks like we are adding an extra layer with no real benefit, as `calypso-build.js` is essentially an alias for `wepback`.

I don't think the [comment in #32630]( https://github.com/Automattic/wp-calypso/pull/32630#issuecomment-491720808) applies anymore, as our package tree has changed a lot since then (we moved to `yarn` and our dependency tree is cleaner). And this change doesn't really solve it, all packages referenced in the config file will still be resolved from `./package.json`, and packages referenced from `calypso-build` will resolve from `packages/calypso-build/package.json`

**Comment by sirreal** [comment]
`calypso-build` was very intentionally designed to be a thin layer over webpack with defaults from our learnings in Calypso. It also includes a `WP` env setting to enable sensible defaults when bundling for WordPress.

It works well enough for a number of projects: Jetpack, Newspack, and a few others.

The motivation for building Calypso with calypso-build is to encourage us to continue sharing what we learn about builds in Calypso with other projects. That's hard to measure, a number of project do seem to use calypso-build, primarily Jetpack, Newspack, and some other smaller Automattic projects. It does not seem to have gained broader traction.

I don't have strong feelings. It's tempting to suggest projects migrate to wp-scripts as the de facto build system for WordPress. `calypso-build` can't really compete with the core provided build system. The projects that need a more robust build system tend to get something set up, requiring customization, and then leave it alone. [Many projects using `calypso-build` seem to be using very old versions](https://github.com/search?l=JSON&q=org%3AAutomattic+calypso-build&type=Code) likely by copying configurations that work.

I'm fine with closing this if it's only going to get in the way of progress 🙂

**Pull request closed without merging**